Sonnen: Enhance #11990 to use `BackupBuffer` instead of switching operating mode

RustyDust opened the pull request:
**Rationale:**
With #11990 the SonnenBatterie will not be drained when loading vehicles anymore (thanks @rivengh !!). However, the current solutions forces a switch of operating modes of the battery between "1" (Manual) and "2" (Auto). This will pose a problem for users who operate their SonnenBatterie in mode "10" (Time of Use - ToU) for example to charge the battery when energy is cheap.

To circumvent this problem my solution uses the `BackupBuffer` functionality of the SonnenBatterie instead. Using the (mislabeled) setting `"EM_USOC"` one can define the minimum charge the SonnenBatterie will hold back when discharging.

Using this settings allows for controlled draining of the SonnenBatterie even during active load plans of evcc.

The current implementation introduces two new config variables:
- `backupbuffer`: The percentage of charge the Sonnenbatterie should reserve during normal operation
- `chargingbuffer`: The percentage of charge the Sonnenbatterie should hold back during charging operations by evcc

**Note:**
It would be nice if instead of fixed config values we could use the current setting of the SonnenBatterie for the `BackupBuffer` as determined on start-up of evcc. Also it would be cool to be able to use the setting `prioritySoc` as stored in the database for the reserved buffer for charging times. I don't know whether it's possible to access those in the template so if it is, a pointer on how to implement this would be great.

RustyDust committed:
Enhance #11990 to use the `BackupBuffer` instead of switching operating mode

**Rationale:**
With #11990 the SonnenBatterie will not be drained when loading vehicles
anymore (thanks @rivengh !!). However, the current solutions forces a switch
of operating modes of the battery between "1" (Manual) and "2" (Auto). This
will pose a problem for users who operate their SonnenBatterie in mode "10"
(Time of Use - ToU) for example to charge the battery when energy is cheap.

To circumvent this problem my solution uses the `BackupBuffer` functionality
of the SonnenBatterie instead. Using the (mislabeled) setting `"EM_USOC"`
one can define the minimum charge the SonnenBatterie will hold back when
discharging.

Using this settings allows for controlled draining of the SonnenBatterie
even during active load plans of evcc.

The current implementation introduces two new config variables:
- `backupbuffer`: The percentage of charge the Sonnenbatterie should
  reserve during normal operation
- `chargingbuffer`: The percentage of charge the Sonnenbatterie should
  hold back during charging operations by evcc

**Note:**
It would be nice if instead of fixed config values we could use the current
setting of the SonnenBatterie for the `BackupBuffer` as determined on
start-up of evcc. Also it would be cool to be able to use the setting
`prioritySoc` as stored in the database for the reserved buffer for charging
times. I don't know whether it's possible to access those in the template
so if it's possible a pointer on how to implement this would be great.

Signed-off-by: Stefan Rubner <stefan@whocares.de>

rivengh commented:
Thanks for picking up my PR.

While I see the intention of your modification, you force to use a non-standard operating mode (`Time of Use`) by default. As mentioned in SonnenBatterie Eco10 user manual:

> "Normalerweise wird der Betriebsmodus Eigenverbrauch verwendet. 
> Die beiden anderen Betriebsmodi Zeitabhängige Steuerung (Time-Of-Use) und Modul-Er-
> weiterung (30%) sind nur in bestimmten Fällen erforderlich und sollten nur von der instal-
> lierenden Elektrofachkraft eingestellt werden."

As I have never used `Time of Use` mode or the backup buffer before, I have the following questions:

1. How about side-effects, when `Time of Use` is selected, but no schedule is configured? Or does it work like mode `self-consumption (2)` then?

2. Are you using a SonnenProtect backup-box in your setup? Is the backup buffer feature also working without it? As it is not accessible via UI:

> Wenn am Speichersystem eine sonnen Notstrombox angeschlossen ist, wird das Feld
> Backup-Reserve angezeigt. Hier können Sie den Notstrompuffer in Prozent auswählen.
> Dieser Anteil der Speicherkapazität wird dann für die Notstromversorgung vorgehalten,
> d. h. dieser Teil der Kapazität steht dem Speichersystem im Normalbetrieb nicht zur Verfü-
> gung.

RustyDust commented:
> While I see the intention of your modification, you force to use a non-standard operating mode (`Time of Use`) by default. 

Well, no, I don't - at least not in within evcc. That's the main reason for the change: using the method in this PR, the mode of the battery won't be changed at all, no matter what the user has set. The use of `BackupBuffer` is independent of the actual operating mode of the SonnenBatterie.

> 1. How about side-effects, when `Time of Use` is selected, but no schedule is configured? Or does it work like mode `self-consumption (2)` then?

Exactly, works as in Mode 2 if the schedule is empty. In fact, when you query the v2 API you should see an empty `EM_ToU_Schedule` already.

> 2. Are you using a SonnenProtect backup-box in your setup? 
Yes.

> Is the backup buffer feature also working without it? 
That I don't know. But it should be easy to test by sending `{"EM_USOC":"xx"}` to the battery in question. If the battery understands the command it should reply with the currently active setting which should be exactly the same string you sent in.

rivengh commented:
> Well, no, I don't - at least not in within evcc. That's the main reason for the change: using the method in this PR, the mode of the battery won't be changed at all, no matter what the user has set. The use of `BackupBuffer` is independent of the actual operating mode of the SonnenBatterie.

Thanks, for clarification. I should have checked your actual modifications before ;)

> That I don't know. But it should be easy to test by sending `{"EM_USOC":"xx"}` to the battery in question. If the battery understands the command it should reply with the currently active setting which should be exactly the same string you sent in.

I'll give it a try when I have some energy in my battery. But sending the value seems to work at a first glance.

RustyDust commented:
> > That's the main reason for the change: using the method in this PR, the mode of the battery won't be changed at all[...]
>
> Thanks, for clarification. I should have checked your actual modifications before ;)

Totally my fault, I should have been clearer in the description of what it does.
Were it only for preventing the SonnenBatterie from getting drained, I'd happily go with your solution. But having a dynamic tariff I try to also load the SonnenBatterie during the "dark months" when electricity is cheap and use it over the day when prices are high. In that scenario resetting the current operating mode of the SonnenBatterie defeats the purpose. And while I agree that most of the SonnenBatterie installations in Germany run in mode 2 the same isn't necessarily true for installations in for example NL and UK where dynamic tariffs seem to be much more common (check discussion here: https://github.com/weltmeyer/ha_sonnenbatterie/issues/31).

> > That I don't know. But it should be easy to test by sending `{"EM_USOC":"xx"}` to the battery in question. If the battery understands the command it should reply with the currently active setting which should be exactly the same string you sent in.
> 
> I'll give it a try, when I have some energy in my battery. But sending the value seems to work at a first glance.

If it doesn't work as expected I guess the easiest solution would be yet another config variable `hasbuffer` depending on which the template would use either your or my way of limiting the drain on the SonnenBatterie.

rivengh commented on templates/definition/meter/sonnenbatterie.yaml:
```diff
@@ -26,6 +26,18 @@ params:
+  - name: backupbuffer
+    required: false
+    help:
+      de: In der Batterie verbleibende Reserve-Kapazität (in %)
+      en: default remaining battery charge (in %)
+    default: 10
```
```suggestion
      de: Verbleibende Reservekapazität in der Batterie (in %)
      en: Remaining reserve capacity in the battery (in %)
    default: 0
    usages: ["battery"]
```
* I suggest using 0 as default, as I assume most users do not own a SonnenProtect backup box. Then a value different from 0 does not make much sense.
* Adding `usages: ["battery"]` is showing this parameter only in the battery documentation.
* Minor rewording of help text

rivengh commented on templates/definition/meter/sonnenbatterie.yaml:
```diff
@@ -26,6 +26,18 @@ params:
+    default: 10
+  - name: chargingbuffer
+    required: false
+    help:
+      de: Maximaler Entladestand der Batterie während des evcc-Ladevorgangs
+      en: Maximum discharge level of the SonnenBatterie during vehicle charging
+    default: 50
```
```suggestion
    help:
      de: Verbleibende Kapazität in der Batterie beim Laden des Fahrzeugs (in %)
      en: Remaining capacity in the battery during vehicle charging (in %)
    default: 100
```
* I suggest using `100` as default in order to have the same behaviour as before (default: do not discharge battery) without configuring additional parameters.
* Adding `usages: ["battery"]` is showing this parameter only in the battery documentation.
* Minor rewording of help text

RustyDust replied:
Using 100 as a default may be dangerous since another user wrote that setting the value higher than the current load of the battery will make the battery start charging. I just tested with my setup and my battery **didn't** start charging. Would be nice if you could verify with yours. Other than that: no objection there.

RustyDust replied:
Ok, I did some more testing and for my setup it is like this:
- if the SonnenBatterie is in mode 1 (Manual) or mode 10 (ToU) setting EM_USOC to a value above the current charge level does nothing
- if, however, the SonnenBatterie is in mode 2 (Auto) setting EM_USOC to a value above the current charge level will make the battery start charging, no matter whether there's enough PV energy available or not.

rivengh replied:
> Ok, I did some more testing and for my setup it is like this:
> 
>     * if the SonnenBatterie is in mode 1 (Manual) or mode 10 (ToU) setting EM_USOC to a value above the current charge level does nothing
> 
>     * if, however, the SonnenBatterie is in mode 2 (Auto) setting EM_USOC to a value above the current charge level will make the battery start charging, no matter whether there's enough PV energy available or not.

Thank you for testing this further. I can reproduce it with my sonnenBatterie eco 8.

From my point of view, the second point disqualifies the use of EM_USOC together with mode 2 (auto/self consumption). We have a good chance of charging not only the car from the grid, but also our home battery!

rivengh reviewed (commented):
Should we consider following information from sonnen APIv2 description for documentation?

>  Possible values for EM_USOC are:
>     0 (disable backup buffer)
>     5 - 100 (backup buffer percentage)

RustyDust committed:
Incorporate suggestions from @rivengh

Signed-off-by: Stefan Rubner <stefan@whocares.de>

RustyDust commented:
Ok, looks like a catch 22 here. If we use the original #11990 version users running their installation not in mode 2 will be upset. If using my proposed version, users with active mode 2 may unintentionally get their SonnenBatterie charged while evcc charges vehicle(s).

The only solution I'm seeing is to really introduce another config variable, my suggestion would be `usingToU`. If set, the ToU specific API calls of my proposal will be used, otherwise the methods of #11990 win.

RustyDust committed:
Use different template paths depending on SB operating mode

This adds yet another config variable `usingtou`. When set to `true`
the EM_USOC type of drain prevention is used. Otherwise the mode
switching variant from @rivengh is used.

This will allow for users to configure the SonnenBatterie to their
current operating mode but has the problem that once the user
changes their decision they'll have to reconfigure evcc as well.

Signed-off-by: Stefan Rubner <stefan@whocares.de>

rivengh commented:
> The only solution I'm seeing is to really introduce another config variable, my suggestion would be `usingToU`. If set, the ToU specific API calls of my proposal will be used, otherwise the methods of #11990 win.

Yeah, that might be a good approach.

I could also think of a more simplistic path without fiddling with EM_USOC:

* A new config parameter - like you said - e.g. `usingToU` or `defaultmode`: to select between `Self Consumption` / `Time Of Use`
* When `Self Consumption` is selected, we change between `Self Consumption` (normal/charge) and `Manual` mode (hold)
* When `Time Of Use` is selected, we change between `Time Of Use` (normal/charge) and `Manual` mode (hold)
 
So users can work with both modes, while keeping configuration simple.

What do you think?

RustyDust commented:
> * When `Time Of Use` is selected, we change between `Time Of Use` (normal/charge) and `Manual` mode (hold)

I agree with the above - except  the last point ;) Since in ToU mode setting of EM_USOC **doesn't** trigger charging the battery it allows for more fine grained control over the charging process. To give you an example:

To cover the "expensive hours" until my PV provides enough energy to fulfil the needs of my home I need approx 10% of the SonnenBattery's total charge. Thus I don't really care whether the charging of the car will drain the battery as long as there are 20% of charge left (10% buffer + 10% to cover the dark hours). So if the battery was at 100% when evcc started charging the car, I'd be perfectly fine with evcc using 80% of the SonnenBattery's capacity to charge the car. Using manual mode I can only disallow discharging the battery completely which would keep the battery fully loaded when in fact I don't need that. That said it'd be totally cool if I could dynamically change the reserve based on solar production forecasts but that has been discussed as a possible enhancement in #10379 and - at least to my understanding - postponed until later.

And before you ask: yes, I'm located in Swabia ;)

As for the naming/logic of the config parameter: I'm totally open there and would use what the evcc devs like best.

In the future it would maybe make sense for evcc to get the state of (battery-) affairs (like mode and other settings that may get modified during charging) before starting to charge and restore the battery to that state once finished. Might come in handy for other logically challenged APIs from other vendors ;)

andig commented:
Sounds more like https://github.com/evcc-io/evcc/pull/10080, maybe another day.

RustyDust commented:
> Sounds more like #10080, maybe another day.

Fwiw: I for one am totally happy without the added functionality proposed there. I can't see any immediate benefit by having the capacity in kWh of my PV battery displayed in evcc. But other users may have other use cases.

rivengh reviewed (changes requested):
I'm sorry it took so long for my review.

I think in general it is a good approach to have both options available: (a) the default Self-Consumption mode and (b) the more advanced Time-of-Use mode that allows the usage of the backup-buffer setting.

My only concern is to rely on the user to set the correct operating mode, when we are not able to verify that.
Therefore I would rather go with the  modification below, which **actively sets the Time-of-Use mode**, when the `timeofuse` setting is enabled.

I have also added some modifications of the documentation.

```patch
diff --git a/templates/definition/meter/sonnenbatterie.yaml b/templates/definition/meter/sonnenbatterie.yaml
index 9ac07487..750eacfa 100644
--- a/templates/definition/meter/sonnenbatterie.yaml
+++ b/templates/definition/meter/sonnenbatterie.yaml
@@ -26,25 +26,25 @@ params:
       de: API Token (benötigt für aktive Batteriesteuerung)
       en: API Token (required for active battery control)
     usages: ["battery"]
-  - name: usingtou
+  - name: timeofuse
     required: false
     help:
-      de: Wird die SonnenBatterie im Modus TimeOfUse betrieben?
-      en: Is the SonnenBattery operated in TimeOfUse mode?
+      de: Nutze den Time-of-Use-Betriebsmodus der SonnenBatterie anstelle des normalerweise verwendeten Eigenverbrauchs-Modus. Nur relevant für aktive Batteriesteuerung.
+      en: Use Time-of-Use operating mode of the SonnenBatterie instead of normally used Self-Consumption mode. Only relevant for active battery control.
     default: false
     usages: ["battery"]
   - name: backupbuffer
     required: false
     help:
-      de: Verbleibende Reservekapazität in der Batterie (in %)
-      en: Remaining reserve capacity in the battery (in %)
+      de: Verbleibende Reservekapazität in der Batterie (in %). Nur relevant für aktive Batteriesteuerung im Time-of-Use-Betriebsmodus.
+      en: Remaining reserve capacity in the battery (in %). Only relevant for active battery control in time-of-use operating mode.
     default: 0
     usages: ["battery"]
   - name: chargingbuffer
     required: false
     help:
-      de: Verbleibende Kapazität in der Batterie beim Laden des Fahrzeugs (in %)
-      en: Remaining capacity in the battery during vehicle charging (in %)
+      de: Verbleibende Kapazität in der Batterie beim Laden des Fahrzeugs (in %). Nur relevant für aktive Batteriesteuerung im Time-of-Use-Betriebsmodus.
+      en: Remaining capacity in the battery during vehicle charging (in %). Only relevant for active battery control in time-of-use operating mode.
     default: 100
     usages: ["battery"]
 render: |
@@ -66,40 +66,70 @@ render: |
     uri: http://{{ .host }}:{{ .port }}/api/v1/status
     jq: .USOC
   {{- if .token }}
-  {{- if eq .usingtou "true" }}
+  {{- if eq .timeofuse "true" }}
   batterymode:
     source: switch
     switch:
     - case: 1 # normal
       set:
-        source: http
-        uri: http://{{ .host }}/api/v2/configurations
-        insecure: true
-        method: PUT
-        headers:
-        - content-type: application/json
-        - Auth-Token: {{ .token }}
-        body: '{"EM_USOC":"{{ .backupbuffer }}"}'  # reset BackupBuffer to default
+        source: sequence
+        set:
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_OperatingMode":"10"}'  # Time-of-Use
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_USOC":"{{ .backupbuffer }}"}'  # default backup buffer
     - case: 2 # hold
       set:
-        source: http
-        uri: http://{{ .host }}/api/v2/configurations
-        insecure: true
-        method: PUT
-        headers:
-        - content-type: application/json
-        - Auth-Token: {{ .token }}
-        body: '{"EM_USOC":"{{ .chargingbuffer }}"}'  # increase BackupBuffer to desired percentage
+        source: sequence
+        set:
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_OperatingMode":"10"}'  # Time-of-Use
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_USOC":"{{ .chargingbuffer }}"}'  # charging backup buffer
     - case: 3 # charge
       set:
-        source: http
-        uri: http://{{ .host }}/api/v2/configurations
-        insecure: true
-        method: PUT
-        headers:
-        - content-type: application/json
-        - Auth-Token: {{ .token }}
-        body: '{"EM_USOC":"{{ .backupbuffer }}"}'  # set BackupBuffer to default
+        source: sequence
+        set:
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_OperatingMode":"10"}'  # Time-of-Use
+        - source: http
+          uri: http://{{ .host }}/api/v2/configurations
+          insecure: true
+          method: PUT
+          headers:
+          - content-type: application/json
+          - Auth-Token: {{ .token }}
+          body: '{"EM_USOC":"{{ .backupbuffer }}"}'  # default backup buffer
   {{- else }}
   batterymode:
     source: switch
```

RustyDust committed:
Merge branch 'master' into sru_sonnen_usoc

RustyDust commented:
> My only concern is to rely on the user to set the correct operating mode, when we are not able to verify that. Therefore I would rather go with the modification below, which **actively sets the Time-of-Use mode**, when the `timeofuse` setting is enabled.
> 

I thought about forcing ToU mode as well but in the end decided against it. The reason for that being that I want evcc to interfere as little as possible with what the user may _knowingly_ be doing. So if the user told evcc during configuration that their SonnenBatterie is running in ToU mode, that's what evcc should expect - but not force. If the user decides to switch the SonnenBatterie to any of the other modes later on I would presume this is done on purpose with awareness of possible side effects.

But seeing that this is more a question of software philosophy than technical necessity, I'd say we ask @andig about his opinion and implement what he prefers how evcc should behave.

andig commented:
Looks like the current implementation works fine and this PR would add additional support for ToU users but also force that mode? That sounds more like a workaround than intended use of ToU.

rivengh commented:
> I thought about forcing ToU mode as well but in the end decided against it. The reason for that being that I wanted evcc to interfere as little as possible with what the user may _knowingly_ be doing. So if the user told evcc during configuration that their SonnenBatterie is running in ToU mode, that's what evcc should expect - but not force. [...]

That sounds totally reasonable and well thought out. Then let's go with it :)

Maybe you can slightly update the docs, to make it clear that the buffer parameters are only applicable in ToU mode (like in my propasal above)?

RustyDust committed:
Amend some of the (ToU) option descriptions

Signed-off-by: Stefan Rubner <stefan@whocares.de>

RustyDust commented:
> Maybe you can slightly update the docs, to make it clear that the buffer parameters are only applicable in ToU mode (like in my propasal above)?

Good point. I made some minor changes since the `backupBuffer` is really only needed with an installed SonnenProtect module.

rivengh reviewed (approved):
LGTM

RustyDust committed:
Typo fix, thx @rivengh

Signed-off-by: Stefan Rubner <stefan@whocares.de>

RustyDust commented on templates/definition/meter/sonnenbatterie.yaml:
```diff
@@ -26,6 +26,27 @@ params:
+    required: false
+    help:
+      de: Eingestellte Reservekapazität der Batterie (in %). Nur relevant bei installiertem SonnenProtect-Modul.
+      en: Configured reserve capacity in the battery (in %). Only relevant with installed SonnenProtect mdule.
```
Fixed.

RustyDust committed:
Merge branch 'master' into sru_sonnen_usoc

andig commented:
Not sure its worth having the additional complexity. More do since evcc itself could in future grid-charge the battery. Maybe rather document that the current implementatation would interfere with ToU?

RustyDust commented:
I agree that this solution adds a bit of complexity and quite frankly I also see this as just some kind of interim solution. But I think that at least for the time being this is better than to interrupt the management of the battery by other software. As a compromise, how about adding the ToU mode as a separate template like it's already done with `sonnenbatterie_eco56.yaml`? This would allow the (assumed) majority of users who are running their SonnenBatterie in mode 2 (auto) to use evcc as is and users who already use ToU mode could switch to the other template if they want to.

If you're thinking about charging PV batteries from the grid using evcc then I think you'd need a completely new internal API anyway that at least allows for passing parameters to the battery related functions. Then, for example with the SonnenBatterie you'd have two different sets of APIs to choose from, both with different advantages and disadvantages. After tinkering with both for a while I'd say the then preferred method would be ToU mainly because it already allows for sending charging plans to the battery. 

That said, if you're planning to expand into charging PV batteries with evcc as well, I'd be really interested in that and I'm also willing to put some time, effort and brain capacity into that.

RustyDust committed:
Merge branch 'master' into sru_sonnen_usoc

RustyDust committed:
Merge branch 'master' into sru_sonnen_usoc

andig commented:
> Maybe rather document that the current implementation would interfere with ToU?

Scheint als würden wir das so nicht umsetzen wollen- closing. Vielen Dank für Eure Mühen! Doku wäre weiter ein Thema.

RustyDust commented:
@andig Tja, schade - und ich bin wirklich traurig. Nicht, weil meine Arbeit umsonst war, das bin ich gewohnt. Sondern weil mit dieser Entscheidung evcc für mich "kaputt" ist. Als ich evcc endeckt habe dachte ich mir: "Endlich jemand der verstanden hat, dass die Leute ihre PV-Installation so nutzen wollen, wie es für sie Sinn macht - und nicht wie Hersteller XYZ denkt, dass es richtig wäre." Und genau das ist jetzt zumindest für mich nicht mehr der Fall. Denn wo mir evcc vorher die Freiheit gegeben hat, mein Equipment so zu konfigurieren, wie **ich** das haben will, muss ich jetzt halt damit leben, dass mir evcc, ohne dass ich es ihm explizit gesagt hätte, in meine Settings reinredet und mir meine PV-Batterie umprogrammiert.

Mir ist schon klar, dass sich evcc im Spannungsfeld zwischen PV-Anlage, Wallbox und Auto-Herstellern bewegt, wo jede der beteiligten Parteien nur ihr jeweiliges Produkt und nicht den Endanwender im Blick hat. Deswegen hatte ich auch versucht, da eine Lösung zu finden, die den Endanwender im Fokus hat - auch wenn das bei der Konfiguration etwas mehr Komplexität erzeugt und wie oben geschrieben sicher auch nicht der Weisheit letzter Schluss ist.

Letztendlich läuft es vermutlich darauf hinaus, dass ihr euch entscheiden müsst, was evcc sein soll: Eine Lösung zum optimierten Laden von "Mobilitäts-Equipment" (E-Autos, E-Bikes, etc.) oder eine Lösung zum möglichst optimierten Laden/Nutzen jedweder, gegebenenfalls konkurrierender Energiespeicher, im Wesentlichen dann eben auch PV-Batterien. 

Die erste Variante ist ziemlich easy, denn da kann man immer mit Daten arbeiten, die in "Echtzeit" vorhanden sind. Habe ich genug PV-Überschuss: ja/nein? Ist der Strom billig genug: ja/nein? Ist der gewünschte Ladestand der "vehicle" erreicht: ja/nein? Alles Daten, die sich in Echtzeit abfragen lassen und mit denen sich auch evcc nahezu in Echtzeit steuern lässt.

Schwierig wird es, wenn die PV-Batterie dazu kommt, weil hier zum einen sofort ein Interessenskonflikt entsteht: Was ist wichtiger, das "vehicle" zu laden oder genug PV-Batteriekapazität zu haben um beispielsweise teure Energiezeiten abzufedern. Zum anderen muss man auf einmal nicht mehr nur mit Echtzeitdaten arbeiten, sondern auch in die nahe Zukunft schauen: Was ist der Strompreis in den nächsten 24h? Was ist die erwartete Leistung der PV-Anlage in den nächsten 24h? Reicht der vorhandene PV-Speicher noch, um teure Zeiten aus der PV-Batterie abzudecken? Und als Nebenfrage: um wie viel weicht der tatsächliche Ertrag der PV von der Vorhersage ab?

In kurz: Das ist der Sprung von jederzeit verfügbaren Live-Daten zu ungenauen und fluktuierenden Wetten auf die Zukunft. Der ist extrem schwierig und wird von der aktuellen API - sorry, den Templates - aktuell gar nicht abgebildet. Zur Lösung des Problems sehe ich zwei Wege:

1) Behandlung von PV-Batterien wie ein "vehicle". Das heisst, die PV-Batterie wird genauso behandelt, bekommt ähnliche Ladepläne (abhängig von Strompreisen und Ertragsvorhersagen) etc. Das bedeutet aber auch, dass für die einzelnen Batterien umfangreichere APIs vorhanden sein müssen als es derzeit mit den Templates der Fall ist um zum Beispiel die PV-Batterie parallel zum "vehicle" zu laden weil der Strom gerade günstig und die PV-Prognose für den nächsten Tag düster ist.

1) Komplettes Ignorieren der PV-Batterie. Sprich: die ist halt da, aber es ist die Aufgabe des Anwenders, die so zu konfigutieren, dass sie eben nicht komplett entladen wird wenn er das nicht will oder eben bevorzugt das Auto geladen wird und dann die PV-Batterie halt leer ist. Am Ende des Tages kann sich jeder Nutzer mit ein wenig Aufwand selbst ausrechnen, was für ihn die (Kosten?-)günstigste Lösung ist.

Oder wieder in kurz: Entweder, evcc managt die PV-Batterie - aber dann richtig -, oder es lässt komplett die Finger weg. Aber eine Lösung in der Art "wir stellen dir halt einfach die Batterie um, ob dir das gefällt oder nicht und erwähnen das in der Doku (die eh keiner liest)" wie jetzt für die SonnenBatterie halte ich für die schlechteste mögliche Variante, da sie genau dem widerspricht, was ich eingangs geschrieben hatte: evcc gibt dem User die Möglichkeit, selbst darüber zu entscheiden, wie sein Equipment funktionieren soll.

Solltest du bis hier gekommen sein: Vielen Dank fürs Lesen. Und bitte nicht falsch verstehen - das ist nicht als Rant gemeint, sondern nur meine Sicht auf das PV-Ökosystem und die Rolle, die evcc aus meiner Warte darin spielt. Das mag sich mit deiner/eurer Sichtweise decken oder nicht, aber es erklärt vielleicht, warum ich traurig bin. Ich finde evcc nach wie vor toll, auch wenn ich jetzt ein wenig mehr Arbeit damit habe um bei neuen Releases meinen Patch reinzufrickeln. Aber das ist ja dann mein Hobby ;)

Wie auch immer: viel Erfolg mit evcc und jetzt bin ich endlich still ;)

Cheers,
-Stefan

RustyDust commented:
> > Sondern weil mit dieser Entscheidung evcc für mich "kaputt" ist. ... Und genau das ist jetzt zumindest für mich nicht mehr der Fall. Denn wo mir evcc vorher die Freiheit gegeben hat, mein Equipment so zu konfigurieren, wie ich das haben will, muss ich jetzt halt damit leben, dass mir evcc, ohne dass ich es ihm explizit gesagt hätte, in meine Settings reinredet und mir meine PV-Batterie umprogrammiert.
> 
> Diese Aussage ist- mit Verlaub- einfach falsch. evcc reicht alle Zustände raus die es benötigt, eine _beliebige_ Steuerung zu implementieren. Nur weil die eingebaute Batteriesteuerung nicht Deinem Anspruch entspricht ist weder etwas kaputt noch wirst Du zu irgendwas gezwungen. Da muss ich doch sehr drum bitten, Emotion und Fakten zu trennen!

Oh, entschuldige, da habe ich mich wohl unklar ausgedrückt. Also am konkreten Beispiel: Ich betreibe meine SonnenBatterie ausschließlich im ToU-Modus (der voreingestellte Mode 2 bietet im Vergleich real eh keine Vorteile, aber das ist eine andere Diskussion) und nutze ein selbst geschriebenes Tool, das sich anhand der Vorhersage des zu erwartenden PV-Ertrags via https://forecast.solar und einer manuell ermittelten erwarteten Abweichung davon, meinem erwarteten Hausverbrauch sowie den Strompreis-Daten von Tibber ausrechnet, ob es für mich günstiger ist, die Batterie zu einer billigen Tarifzeit aus dem Netz ein wenig aufzufüllen um teure Phasen zu überbrücken, für die ich nach der Prognose nicht genug Batterie-Kapazität hätte (Betrifft i.W. die Morgenstunden oder halt den Winter). Letztlich erstellt das Tool dann einen Ladeplan, der sehr ähnlich wie der aktuelle von evcc für Vehicles ist und schickt das an die SonnenBatterie, die dann automatisch in der festgelegten Zeitspanne aus dem Netz lädt **wenn sie im ToU-Modus ist**. _In allen anderen Modi werden diese Ladepläne ignoriert._

Seit der Änderung des SonnenBatterie-Templates durch #11990 passiert nun Folgendes: Wird während oder irgenwann vor den Zeitfenstern, in denen die SonnenBatterie geladen werden soll, auch mein Vehicle geladen (was extrem wahrscheinlich ist, da das Auto aktuell unter Tags meistens unterwegs und dank Ladeplan die "günstigen Zeitfenster" halt identisch sind), dann schaltet mir das Template den ToU-Modus **aus**. Weil es eben nur zwischen Mode 1 (Manual) und Mode 2 (Auto) wechselt. Sprich: evcc killt mir dann meinen Ladeplan. Ein weiterer Effekt ist, dass jetzt nach jedem Laden die SonnenBatterie anschließend im Mode 2 (Auto) verbleibt, den ich halt so nicht eingestellt hatte.

Und das hatte ich halt als "reinreden" in meine Settings und beschneiden meiner Möglichkeiten bezeichnet, mein Equipment so zu konfigurieren, wie ich es gerne hätte. Denn evcc passt da meine SonneBatterie eben so an, wie evcc und nicht wie ich das gerne hätte. Hat mit Emotionen aber nix zu tun, ist halt so ;)

Da ich aber ganz klar auch den Wert von #11990 für Nutzer gesehen habe, die standardmäßig in Mode 2 arbeiten, hatte ich versucht, das Template so zu erweitern, dass es für _alle_ Betriebsmodi der SonnenBatterie funktioniert - ja, mit ein wenig mehr Komplexität bei der Konfiguration. Und es hat mich ein wenig traurig gemacht, dass das anscheinend nicht als Vorteil/Verbesserung gesehen wurde und eben evcc für mich / meinen Anwendungsfall "kaputt" ist. Und ja, wahrscheinlich war "kaputt" ein zu hartes Wort, vielleicht hätte ich "beschädigt" sagen sollen, sorry dafür.

Natürlich wäre es noch schöner, wenn evcc vorher den aktuellen Betriebsmodus der Batterie abfragt und dann intern das passende Verfahren zum Verhindern des Battery-Drains beim Laden des Vehicles wählen würde. Aber da sind wir wieder bei APIs und so und ja eh einer Meinung, dass die Templates für sowas nicht ausreichen.

andig commented:
> Und das hatte ich halt als "reinreden" in meine Settings und beschneiden meiner Möglichkeiten bezeichnet, mein Equipment so zu konfigurieren, wie ich es gerne hätte. Denn evcc passt da meine SonneBatterie eben so an, wie evcc und nicht wie ich das gerne hätte. Hat mit Emotionen aber nix zu tun, ist halt so ;)

@RustyDust genau diesen Punkt sehe ich anders. Es gibt keinen Zwang, die Battery Control zu aktivieren. Du kannst das komplett in HA tun ohne dass Dir evcc dazwischen funkt! Die Freiheit hast Du selbstverständlich auch weiter...

RustyDust commented:
Ok, point taken. Dann verliere ich halt Genauigkeit bei der Berechnung der Ladekosten, weil evcc dann nicht mehr mitkriegt, wie viel aus der Batterie kam. Aber damit kann ich leben. Danke, hatte tatsächlich übersehen, dass ich die Batterie gar nicht definieren muss, sondern die immer in Verbindung mit der PV gesehen, weil selbe API und so.

premultiply commented:
Du brauchst einfach nur nicht die Batteriekontrolle zu aktivieren!

andig commented:
Genau. Einfach nicht so kompliziert machen.

andig commented:
> Danke, hatte tatsächlich übersehen, dass ich die Batterie gar nicht definieren muss, sondern die immer in Verbindung mit der PV gesehen, weil selbe API und so.

Ich glaube da geht es immer noch durcheinander. Du kannst die Batterie sehr wohl definieren, muss aber die aktivie Batteriesteuerung ja nicht einschalten.

> Dann verliere ich halt Genauigkeit bei der Berechnung der Ladekosten, weil evcc dann nicht mehr mitkriegt, wie viel aus der Batterie kam. Aber damit kann ich leben. 

Solang evcc die Batterieleistung sieht kann es auch die Kostenberechnung korrekt machen. Ob es das selbst steuert oder nicht.

RustyDust commented:
> Ich glaube da geht es immer noch durcheinander. Du kannst die Batterie sehr wohl definieren, muss aber die aktive Batteriesteuerung ja nicht einschalten.

Keine Sorge, das ist genau das, was ich gemacht habe.
